site/app/libraries/MainRouter.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+
+namespace app\libraries;
+
+use Symfony\Component\Config\FileLocator;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Routing\RequestContext;
+use Symfony\Component\Routing\Matcher\UrlMatcher;
+use Symfony\Component\Routing\Loader\AnnotationDirectoryLoader;
+use Doctrine\Common\Annotations\AnnotationReader;
+
+
+class MainRouter {
+    /** @var Core  */
+    protected $core;
+
+    /** @var Request  */
+    protected $request;
+
+    /** @var bool */
+    protected $logged_in;
+
+    /** @var array */
+    protected $parameters;
+
+    /** @var UrlMatcher  */
+    protected $matcher;
+
+    public function __construct(Core $core, $logged_in) {
+        $this->core = $core;
+        $this->request = Request::createFromGlobals();
+        $this->logged_in = $logged_in;
+
+        $fileLocator = new FileLocator();
+        $annotationLoader = new AnnotatedRouteLoader(new AnnotationReader());
+        $loader = new AnnotationDirectoryLoader($fileLocator, $annotationLoader);
+        $collection = $loader->load(realpath(__DIR__ . "/../controllers"));
+
+        $this->matcher = new UrlMatcher($collection, new RequestContext());
+        $this->parameters = $this->matcher->matchRequest($this->request);
+    }
+
+    public function run() {
+        $this->loginCheck();
+
+        $controllerName = $this->parameters['_controller'];
+        $methodName = $this->parameters['_method'];
+
+        if (in_array('semester', $this->parameters) && in_array('course', $this->parameters)) {
+            $semester = $this->parameters['semester'];
+            $course = $this->parameters['course'];
+            $this->core->loadConfig($semester, $course);
+        }
+
+        $controller = new $controllerName($this->core);
+        return $controller->$methodName();
```

Comment:
how could we pass in the potential request args for a given path to the function?

e.g. If we have `/submission/{gradeable_id}`, how would we pass that `gradeable_id` to the function: `$controller->$methodName($gradeable_id);`

Reply:
Thank you very much for reminding me of this. I did not notice this because `$course` and `$semester` are quite special that they are processed before calling the navigation controller. I have fixed this by using `call_user_func_array` and removing the parameters starting with `_`

Reply:
This screenshot shows the expected behavior.
![Screenshot from 2019-06-04 13-53-06](https://user-images.githubusercontent.com/20266703/58854503-1e3acb80-86d0-11e9-843c-4864a15f6ada.png)

Reply:
I changed the route to `/{_semester}/{_course}` in the latest commit considering that `$semester` and `$course` are hardly used in a specific function. (Course-related data are loaded before entering a `controller::method`.)

The most recent commit allows `http://192.168.56.111/s19/sample/student/grading_homework`.

The function looks like:
```
/**
 * @Route("/{_semester}/{_course}/student/{gradeable_id}")
 * @return array
 */
public function showHomeworkPage($gradeable_id = null) 
```

I would use the following code to ensure backward compatibility before the new router is fully set up.

```
if (!isset($gradeable_id)) {
    $gradeable_id = (isset($_REQUEST['gradeable_id'])) ? $_REQUEST['gradeable_id'] : null;
}
```